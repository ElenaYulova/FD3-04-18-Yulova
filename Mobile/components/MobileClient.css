.MobileClient
{
  padding: 2px 0;
}

.MobileClientBalance
{
  float: right;
  width: 30px;
  font-weight: bold;
  text-align: right;
}


.MobileClientActiveActive {
  color: green;
  font-weight: bold;
}

.MobileClientActiveBlocked {
  color: red;
  font-weight: bold;
}