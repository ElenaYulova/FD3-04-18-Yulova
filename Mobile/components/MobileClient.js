import React from 'react';
import PropTypes from 'prop-types';

import './MobileClient.css';

class MobileClient extends React.PureComponent {

  static propTypes = {
    client: PropTypes.shape({
      id: PropTypes.number.isRequired,
      fam: PropTypes.string.isRequired,
      im: PropTypes.string.isRequired,
      otch: PropTypes.string.isRequired,
      balance: PropTypes.number.isRequired,
      shown: PropTypes.bool.isRequired,
      edited: PropTypes.bool,
    }),
       
  };
  static defaultProps = {
    client: {
      edited: false,
    }
  }
  state = {
    FIO: {
      fam: this.props.client.fam,
      im: this.props.client.im,
      otch: this.props.client.otch,      
    },
    balance: this.props.balance,
    edited: this.props.client.edited,
    activeClient: this.props.client.balance >= 0,
  };

  componentWillReceiveProps = (newProps) => {
    //console.log("MobileClient id="+this.props.id+" componentWillReceiveProps");
    var newClient = newProps.client;
    if (this.state.FIO.fam != newClient.fam || this.state.FIO.im != newClient.im || this.state.FIO.otch != newClient.otch) {
      this.setState({
        FIO: {
          fam: newClient.fam,
          im: newClient.im,
          otch: newClient.otch,
        },
        edited: false,
      });
    }
    if (this.state.balance != newClient.balance) {
      this.setState({
        balance: newProps.client.balance,
        isActive: newProps.client.balance >= 0,
      });
    }
  };
  editButton = () => {
    this.setState({ edited: true });
  }
  increaseBalance = () => {
    this.props.onBalanceChange(this.props.client.id, ++this.state.balance);
    if (this.state.balance >= 0) {
      this.setState({
        activeClient: true,
      });
    }
  }

  decreaseBalance = () => {
    this.props.onBalanceChange(this.props.client.id, --this.state.balance);
    if (this.state.balance < 0) {
      this.setState({
        activeClient: false,
      });
    }
  }
  deleteClient = () => {
    this.props.onDelete(this.props.client.id);
  }
  saveForm = (EO) => {
    EO.preventDefault();
    var clientForm = EO.target;
    var FIO = {
      fam: clientForm.fam.value,
      im: clientForm.im.value,
      otch: clientForm.otch.value,
    };
    this.props.onFIOChange(this.props.client.id, FIO);
    this.setState({
       edited: false 
      });
  }
  render() {

    console.log("MobileClient id="+this.props.id+" render");
    
    return (
      <div className='MobileClient'>{
        this.props.client.shown && <div className='MobileClient'>
            <span className="MobileClientId">{this.props.client.id}</span>
            <div className='MobileClientFio'>
              {
                this.state.edited
                  ?
                  <form onSubmit={this.saveForm} className="MobileClientFioForm">
                    <input type="text" defaultValue={this.state.fam} name="fam" />
                    <input type="text" defaultValue={this.state.im} name="im" />
                    <input type="text" defaultValue={this.state.otch} name="otch" />
                    <input type="submit" value="Save" />
                  </form>
                  :
                  <div className="MobileClientFioContainer">
                    <span>{this.state.fam + " " + this.state.im + " " + this.state.otch}</span>
                    <button onClick={this.editButton}>Edit</button>
                    </div>
              }
              </div>
              <div>
              <span className="MobileClientBalance">{this.state.balance}</span>
              <button onClick={this.decreaseBalance}>Decrease</button>
              <button onClick={this.increaseBalance}>Increase</button>
              </div>
              <div className="MobileClientActive">
              {
                this.state.activeClient
                  ?
                  <span className="MobileClientActiveActive">Active</span>
                  :
                  <span className="MobileClientActiveBlocked">Blocked</span>
              }
            </div>
            <div className="MobileClientDelete">
              <button onClick={this.deleteClient}>Delete</button>
            </div>
        </div>
      }
       </div> 
        
      
    );

  }

}

export default MobileClient;
