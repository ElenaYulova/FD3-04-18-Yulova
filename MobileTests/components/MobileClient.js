import React from 'react';
import PropTypes from 'prop-types';

import './MobileClient.css';

class MobileClient extends React.PureComponent {

  static propTypes = {
    client: PropTypes.shape({
      id: PropTypes.number.isRequired,
      fam: PropTypes.string.isRequired,
      im: PropTypes.string.isRequired,
      otch: PropTypes.string.isRequired,
      balance: PropTypes.number.isRequired,
      edited: PropTypes.bool.isRequired,
    }),
    onFIOChange: PropTypes.func.isRequired,
    onBalanceChange: PropTypes.func.isRequired,
    onDelete: PropTypes.func.isRequired,  
  };
  
  state = {
    
      fam: this.props.client.fam,
      im: this.props.client.im,
      otch: this.props.client.otch,      
    
    balance: this.props.client.balance,
    edited: this.props.client.edited,
    active: this.props.client.balance >= 0,
  };

  componentWillReceiveProps = (newProps) => {
    //console.log("MobileClient id="+this.props.id+" componentWillReceiveProps");
    var fam = this.state.fam;
    var im = this.state.im;
    var otch = this.state.otch;
    var newClient = newProps.client;
    if (this.fam != newClient.fam || im != newClient.im || otch != newClient.otch) {
      this.setState({
        
          fam: newClient.fam,
          im: newClient.im,
          otch: newClient.otch,
        
        edited: false,
      });
    }
    if (this.state.balance != newClient.balance) {
      this.setState({
        balance: newProps.client.balance,
        active: newProps.client.balance >= 0,
      });
    }
    if (this.state.edited != newClient.edited) {
      this.setState({
        edited: newClient.edited,
      });
    }
  };
  editButton = () => {
    this.setState({ edited: true });
  }
  increaseBalance = () => {
    this.props.onBalanceChange(this.props.client.id, ++this.state.balance);
    if (this.state.balance >= 0) {
      this.setState({
        active: true,
      });
      // console.log(this.state.balance)
    }
  }

  decreaseBalance = () => {
    this.props.onBalanceChange(this.props.client.id, --this.state.balance);
    if (this.state.balance < 0) {
      this.setState({
        active: false,
      });
    }
  }
  deleteClient = () => {
    this.props.onDelete(this.props.client.id);
  }
  saveForm = (EO) => {
    EO.preventDefault();
    var clientForm = EO.target;
    var FIO = {
      fam: clientForm.fam.value,
      im: clientForm.im.value,
      otch: clientForm.otch.value,
    };
    this.props.onFIOChange(this.props.client.id, FIO);
    
  }
  render() {

    console.log("MobileClient id="+this.props.client.id+" render");
    
    return (
      <div className='MobileClient'>{
        <div className='MobileClient'>
            <span className="MobileClientId">{this.props.client.id}</span>
            <div className='MobileClientFio'>
              {
                this.state.edited
                  ?
                  <form onSubmit={this.saveForm} className="MobileClientFioForm">
                    <input type="text" defaultValue={this.state.fam} name="fam" />
                    <input type="text" defaultValue={this.state.im} name="im" />
                    <input type="text" defaultValue={this.state.otch} name="otch" />
                    <input type="submit" value="Save" />
                  </form>
                  :
                  <div className="MobileClientFioContainer">
                    <span>{this.state.fam + " " + this.state.im + " " + this.state.otch}</span>
                    <button onClick={this.editButton}>Edit</button>
                    </div>
              }
              </div>
              <div>
              <span className="MobileClientBalance">{this.state.balance}</span>
              <button className="decrease" onClick={this.decreaseBalance}>Decrease</button>
              <button className="increase" onClick={this.increaseBalance}>Increase</button>
              </div>
              <div className="MobileClientActive">
              {
                this.state.active
                  ?
                  <span className="MobileClientActiveActive">Active</span>
                  :
                  <span className="MobileClientActiveBlocked">Blocked</span>
              }
            </div>
            <div className="MobileClientDelete">
              <button className="delete"onClick={this.deleteClient}>Delete</button>
            </div>
        </div>
      }
       </div> 
        
      
    );

  }

}

export default MobileClient;
