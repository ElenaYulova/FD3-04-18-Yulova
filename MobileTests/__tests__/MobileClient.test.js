"use strict";

import React from 'react';
import renderer from 'react-test-renderer';

import {shallow, configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import MobileClient from '../components/MobileClient';

configure({ adapter: new Adapter() });

describe('>>>MobileClient --- Shallow Render REACT COMPONENTS',()=>{
//shapshots
   const client1 = {
    id: 1,
    fam: 'fam1',
    im: 'im1',
    otch: 'otch1',
    balance: 0,
    edited: false,
  };
  const client2 = {
    id: 2,
    fam: 'fam2',
    im: 'im2',
    otch: 'otch2',
    balance: 1,
    edited: true,
  };
  const client3 = {
    id: 3,
    fam: 'fam3',
    im: 'im3',
    otch: 'otch3',
    balance: -1,
    edited: false,
  };
  const client1Component = <MobileClient 
  client={client1}
  onFIOChange={() => {}}
  onBalanceChange={() => {}}
  onDelete={() => {}}
   />;

   const client2Component = <MobileClient 
   client={client2}
   onFIOChange={() => {}}
   onBalanceChange={() => {}}
   onDelete={() => {}}
    />;

    const client3Component = <MobileClient 
    client={client3}
    onFIOChange={() => {}}
    onBalanceChange={() => {}}
    onDelete={() => {}}
     />;

     const component1 = renderer.create(client1Component);

     it('renders usual client', () => {
       let componentTree = component1.toJSON();
       expect(componentTree).toMatchSnapshot();
     });

     const component2 = renderer.create(client2Component);

    it('renders edited client', () => {
        let componentTree = component2.toJSON();
        expect(componentTree).toMatchSnapshot();
    });

    const component3 = renderer.create(client3Component);

    it('renders blocked client', () => {
        let componentTree = component3.toJSON();
        expect(componentTree).toMatchSnapshot();
    });

    //onDelete
    it('onDelete works', () => {
        const onDelete = jest.fn();
        const wrapper = shallow(
          <MobileClient client={client1} onDelete={onDelete}
            onFIOChange={() => { }}
            onBalanceChange={() => { }} />
        );
    
        wrapper.find('.delete').simulate('click');
    
        expect(onDelete).toBeCalledWith(client1.id);
      });
//onBalanceChange
    it('onBalanceChange works', () => {
        const onBalanceChange = jest.fn();
        const wrapper = shallow(
          <MobileClient client={client1} onDelete={() => { }}
            onFIOChange={() => { }}
            onBalanceChange={onBalanceChange} />
        );
    
        wrapper.find('.increase').simulate('click');
    
        expect(onBalanceChange).toBeCalledWith(client1.id, 1);
      });
    
      it('onBalanceChange works', () => {
        const onBalanceChange = jest.fn();
        const wrapper = shallow(
          <MobileClient client={client1} onDelete={() => { }}
            onFIOChange={() => { }}
            onBalanceChange={onBalanceChange} />
        );
    
        wrapper.find('.decrease').simulate('click');
    
        expect(onBalanceChange).toBeCalledWith(client1.id, -1);
      });

});
