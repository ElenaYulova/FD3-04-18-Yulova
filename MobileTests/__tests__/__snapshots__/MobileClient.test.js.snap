// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>>MobileClient --- Shallow Render REACT COMPONENTS renders blocked client 1`] = `
<div
  className="MobileClient"
>
  <div
    className="MobileClient"
  >
    <span
      className="MobileClientId"
    >
      3
    </span>
    <div
      className="MobileClientFio"
    >
      <div
        className="MobileClientFioContainer"
      >
        <span>
          fam3 im3 otch3
        </span>
        <button
          onClick={[Function]}
        >
          Edit
        </button>
      </div>
    </div>
    <div>
      <span
        className="MobileClientBalance"
      >
        -1
      </span>
      <button
        className="decrease"
        onClick={[Function]}
      >
        Decrease
      </button>
      <button
        className="increase"
        onClick={[Function]}
      >
        Increase
      </button>
    </div>
    <div
      className="MobileClientActive"
    >
      <span
        className="MobileClientActiveBlocked"
      >
        Blocked
      </span>
    </div>
    <div
      className="MobileClientDelete"
    >
      <button
        className="delete"
        onClick={[Function]}
      >
        Delete
      </button>
    </div>
  </div>
</div>
`;

exports[`>>>MobileClient --- Shallow Render REACT COMPONENTS renders edited client 1`] = `
<div
  className="MobileClient"
>
  <div
    className="MobileClient"
  >
    <span
      className="MobileClientId"
    >
      2
    </span>
    <div
      className="MobileClientFio"
    >
      <form
        className="MobileClientFioForm"
        onSubmit={[Function]}
      >
        <input
          defaultValue="fam2"
          name="fam"
          type="text"
        />
        <input
          defaultValue="im2"
          name="im"
          type="text"
        />
        <input
          defaultValue="otch2"
          name="otch"
          type="text"
        />
        <input
          type="submit"
          value="Save"
        />
      </form>
    </div>
    <div>
      <span
        className="MobileClientBalance"
      >
        1
      </span>
      <button
        className="decrease"
        onClick={[Function]}
      >
        Decrease
      </button>
      <button
        className="increase"
        onClick={[Function]}
      >
        Increase
      </button>
    </div>
    <div
      className="MobileClientActive"
    >
      <span
        className="MobileClientActiveActive"
      >
        Active
      </span>
    </div>
    <div
      className="MobileClientDelete"
    >
      <button
        className="delete"
        onClick={[Function]}
      >
        Delete
      </button>
    </div>
  </div>
</div>
`;

exports[`>>>MobileClient --- Shallow Render REACT COMPONENTS renders usual client 1`] = `
<div
  className="MobileClient"
>
  <div
    className="MobileClient"
  >
    <span
      className="MobileClientId"
    >
      1
    </span>
    <div
      className="MobileClientFio"
    >
      <div
        className="MobileClientFioContainer"
      >
        <span>
          fam1 im1 otch1
        </span>
        <button
          onClick={[Function]}
        >
          Edit
        </button>
      </div>
    </div>
    <div>
      <span
        className="MobileClientBalance"
      >
        0
      </span>
      <button
        className="decrease"
        onClick={[Function]}
      >
        Decrease
      </button>
      <button
        className="increase"
        onClick={[Function]}
      >
        Increase
      </button>
    </div>
    <div
      className="MobileClientActive"
    >
      <span
        className="MobileClientActiveActive"
      >
        Active
      </span>
    </div>
    <div
      className="MobileClientDelete"
    >
      <button
        className="delete"
        onClick={[Function]}
      >
        Delete
      </button>
    </div>
  </div>
</div>
`;
